### PR TITLE
CIoU Loss

### PR DESCRIPTION
**Pull request by IMvision12**
# What does this PR do?

Fixes: #1836

@ianstenbit

### CODE REVIEW

File: keras_cv/losses/ciou_loss.py [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+        (5, 10, 5),
+        minval=0,
+        maxval=10,
+        dtype=tf.dtypes.int32)
+    y_pred = tf.random.uniform(
+        (5, 10, 4),
+        minval=0,
+        maxval=10,
+        dtype=tf.dtypes.int32)
+    loss = CIoULoss()
```
Comment (ianstenbit) [inline keras_cv/losses/ciou_loss.py:53]: `keras_cv.losses.CIoULoss()`

File: keras_cv/losses/ciou_loss.py [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+    bounding box coordinates but also considers the aspect ratio and center
+    distance of the boxes. The length of the last dimension should be 4 to
+    represent the bounding boxes.
+
+    Args:
+        bounding_box_format: a case-insensitive string (for example, "xyxy").
+            Each bounding box is defined by these 4 values. For detailed
+            information on the supported formats, see the [KerasCV bounding box
+            documentation](https://keras.io/api/keras_cv/bounding_box/formats/).
+        axis: the axis along which to mean the ious, defaults to -1.
```
Comment (ianstenbit) [inline keras_cv/losses/ciou_loss.py:35]: nit: s/ious/IoUs

File: keras_cv/losses/ciou_loss.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
+# Copyright 2022 The KerasCV Authors
```
Comment (ianstenbit) [inline keras_cv/losses/ciou_loss.py:1]: 2023

**Review by ianstenbit — CHANGES_REQUESTED** [review]
Nice -- Thank you for the great PR!

Let's add some unit tests including a numerical test.

Either in this PR or in a follow-up, we should also delete the `YOLOV8IoULoss` and use this one there intead.

**Comment by IMvision12** [comment]
Yeah, I will create a follow-up PR for updating yolov8 loss!

**Comment by IMvision12** [comment]
hey @ianstenbit do I need to add this for supporting different formats of bbox ? 

```python
def compute_ciou(self, boxes1, boxes2):
        target_format = "xywh"
        if bounding_box.is_relative(self.bounding_box_format):
            target_format = bounding_box.as_relative(target_format)

        boxes1 = bounding_box.convert_format(
            boxes1, source=self.bounding_box_format, target=target_format
        )

        boxes2 = bounding_box.convert_format(
            boxes2, source=self.bounding_box_format, target=target_format
        )
```

**Comment by IMvision12** [comment]
@ianstenbit is there anything missing in `ciou_loss.py`, coz I am getting some shape errors while testing?

File: keras_cv/losses/ciou_loss.py [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+    """Implements the Complete IoU (CIoU) Loss
+
+    CIoU loss is an extension of GIoU loss, which further improves the IoU
+    optimization for object detection. CIoU loss not only penalizes the
+    bounding box coordinates but also considers the aspect ratio and center
+    distance of the boxes. The length of the last dimension should be 4 to
+    represent the bounding boxes.
+
+    Args:
+        bounding_box_format: a case-insensitive string (for example, "xyxy").
```
Comment (ianstenbit) [inline keras_cv/losses/ciou_loss.py:31]: We'll need to use this parameter and convert boxes to our compute format `xyxy`

**Review by ianstenbit — COMMENTED** [review]
Can you add some more info about the shape issues you're getting?

**Comment by IMvision12** [comment]
@ianstenbit 

The test is failing for `test_output_shape_reduction_none`, I printed the shape of `ciou_loss(y_true, y_pred).shape` : 

```python
y_true = tf.random.uniform(shape=(2, 2, 4), minval=0, maxval=10, dtype=tf.int32)
y_pred = tf.random.uniform(shape=(2, 2, 4), minval=0, maxval=20, dtype=tf.int32)
ciou_loss = CIoULoss(bounding_box_format="xyxy", reduction="none")
print(ciou_loss(y_true, y_pred).shape)
```

Output = (2, 2, 1)

File: keras_cv/losses/ciou_loss.py [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+
+        if y_true.shape[-2] != y_pred.shape[-2]:
+            raise ValueError(
+                "CIoULoss expects number of boxes in y_pred to be equal to the "
+                "number of boxes in y_true. Received number of boxes in "
+                f"y_true={y_true.shape[-2]} and number of boxes in "
+                f"y_pred={y_pred.shape[-2]}."
+            )
+
+        ciou = self.compute_ciou(y_true, y_pred)
```
Comment (ianstenbit) [inline keras_cv/losses/ciou_loss.py:143]: We probably just need to squeeze this on axis=-1

`ciou = tf.squeeze(self.compute_ciou(y_true, y_pred), axis=-1)`

File: keras_cv/losses/ciou_loss_test.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+        )
+
+        ciou_loss = CIoULoss(bounding_box_format="xyxy", reduction="none")
+
+        self.assertAllEqual(
+            [2, 2],
+            ciou_loss(y_true, y_pred).shape,
+        )
+
+    def test_output_shape_relative_formats(self):
```
Comment (ianstenbit) [inline keras_cv/losses/ciou_loss_test.py:48]: I'm not sure the output shape assertion is particularly meaningful here, just testing a relative format, right?

So maybe we should parameterize the `output_value` test to run as both absolute and relative format, then we can assert that the losses match

(e.g. scale the y_true and y_pred by 1/640 or similar, then say format = "rel_xyxy" and assert the value is the same)

**Review by ianstenbit — APPROVED** [review]
Looks good after one small change -- thanks!

File: keras_cv/losses/ciou_loss_test.py [hunk 7]
```diff
@@ -65,21 +66,38 @@ def test_output_shape_relative_formats(self):
...
-            [0, 0, 7, 3],
-            [4, 5, 5, 6],
-            [2, 1, 3, 3],
-        ]
-
-        ciou_loss = CIoULoss(bounding_box_format="xyxy")
-
-        # expected value for these values is 1.0320253
-        self.assertAllClose(ciou_loss(y_true, y_pred), 1.0320253)
+        def test_loss_format(bounding_box_format):
```
Comment (ianstenbit) [inline keras_cv/losses/ciou_loss_test.py:69]: Instead of making a sub-function, let's use a `Parameterized` test (see example [here](https://github.com/keras-team/keras-cv/blob/master/keras_cv/models/object_detection/yolo_v8/yolo_v8_detector_test.py#L113)) for this

File: keras_cv/losses/ciou_loss_test.py [hunk 8]
```diff
@@ -65,21 +66,38 @@ def test_output_shape_relative_formats(self):
...
+                [0, 0, 5, 6],
+                [0, 0, 7, 3],
+                [4, 5, 5, 6],
+                [2, 1, 3, 3],
+            ]
+
+            if bounding_box_format == "xyxy":
+                ciou_loss = CIoULoss(bounding_box_format="xyxy")
+                # expected value is 1.032025
+                expected_loss = 1.032025
```
Comment (ianstenbit) [inline keras_cv/losses/ciou_loss_test.py:87]: For these two, we can just specify `expected_loss = 1.03202` for both versions, as the difference is just a tiny numerical error.

If the test fails for one version, we can specify `atol=0.0001` in the `assertAllClose`

**Review by ianstenbit — APPROVED** [review]
Thanks! Looks good -- just want to clean up the test parameterization a bit

**Review by ianstenbit — APPROVED** [review]
Thanks for the PR and the revisions!

**Comment by NyxAither** [comment]
Not sure if this was a deliberate decision (or I'm just dumb and misunderstanding the code) but FYI it appears to me that the keras compute_ciou method omits the conditional handling of the trade-off parameter alpha from the original implementation in the cited CIoU paper (screenshot from paper). Thanks!
image
